OpenRA.Mods.Common/Widgets/Logic/IrcLogic.cs

```diff
@@ -0,0 +1,364 @@
+			widget.Bounds.Height = font.Measure(item).Y;
+			widget.GetText = () => item;
+			widget.Id = item;
+			return widget;
+		}
+
+		void AddChatLine(string text)
+		{
+			Log.Write("irc", text);
+			Game.RunAfterTick(() =>
+			{
+				var scrolledToBottom = historyPanel.ScrolledToBottom;
+
+				var newChild = MakeLabelWidget(historyTemplate, text);
+				historyPanel.AddChild(newChild);
+
+				if (scrolledToBottom)
+					historyPanel.ScrollToBottom(smooth: true);
+			});
+		}
+
+		bool EnterPressed()
+		{
+			if (!inputBox.Text.Any())
+				return true;
+
+			var text = inputBox.Text;
+			inputBox.Text = "";
+
+			if (text.StartsWith("/nick "))
+			{
+				var nick = text.Replace("/nick ", string.Empty);
+				if (Rfc2812.IsValidNickname(nick))
+					Irc.Client.RfcNick(nick);
+				else
+					AddChatLine("Invalid nickname.");
+			}
+			else if (text.StartsWith("/ping "))
+			{
+				Irc.Client.RfcPing(Irc.Client.GetIrcUser(text.Replace("/ping ", string.Empty)).Host);
+				pingSent = true;
+			}
+			else if (text.StartsWith("/"))
+				AddChatLine("Unknown command.");
+			else
+			{
+				AddChatLine("[{0}] <{1}> {2}".F(DateTime.Now.ToString(Game.Settings.Irc.TimestampFormat), Irc.Client.Nickname, text));
+				Irc.Client.SendMessage(SendType.Message, "#" + Game.Settings.Irc.Channel, text);
+			}
+
+			return true;
+		}
+
+		void OnConnected(object sender, EventArgs e)
+		{
+			AddChatLine("Connected.");
+
+			if (!Rfc2812.IsValidNickname(Game.Settings.Irc.Nickname))
+			{
+				AddChatLine("Invalid nickname. Can't login.");
+				return;
+			}
+
+			Irc.Client.Login(new[] { Game.Settings.Irc.Nickname }, "in-game IRC client", 0, "OpenRA");
+
+			Irc.Client.RfcJoin("#" + Game.Settings.Irc.Channel);
+		}
+
+		void OnError(object sender, ErrorEventArgs e)
+		{
+			AddChatLine("Error: ".F(e.ErrorMessage));
+			Log.Write("irc", "Error: {0}".F(e.ErrorMessage));
+		}
+
+		void OnRawMessage(object sender, IrcEventArgs e)
+		{
+			Log.Write("irc", e.Data.RawMessage);
+		}
+
+		void OnChannelMessage(object sender, IrcEventArgs e)
+		{
+			AddChatLine("[{0}] <{1}> {2}".F(DateTime.Now.ToString(Game.Settings.Irc.TimestampFormat), e.Data.Nick, e.Data.Message));
+		}
+
+		void OnJoin(object sender, JoinEventArgs e)
+		{
+			AddChatLine("{0} joined channel {1}.".F(e.Who, e.Channel));
+			channel = Irc.Client.GetChannel(e.Channel);
+			SyncNicknamePanel();
+		}
+
+		void OnChannelActiveSynced(object sender, IrcEventArgs e)
+		{
+			channel = Irc.Client.GetChannel(e.Data.Channel);
+
+			AddChatLine("{0} users online".F(channel.Users.Count));
+
+			if (!string.IsNullOrEmpty(channel.Topic))
+				AddChatLine("*** Topic: {0}".F(channel.Topic));
+
+			SyncNicknamePanel();
+		}
+
+		void OnNickChange(object sender, NickChangeEventArgs e)
+		{
+			AddChatLine("{0} is now known as {1}.".F(e.OldNickname, e.NewNickname));
+			SyncNicknamePanel();
+		}
+
+		void SyncNicknamePanel()
+		{
+			if (channel == null)
+				return;
+
+			var users = channel.Users;
```

Comment:
Sorry I meant: do we need to copy the whole collection in order to do this safely, rather than just caching the variable?

What happens if the IRC thread changes the collection whilst we're enumerating it, for example?